include/aspect/particle/property/cpo_bingham_average.h

```diff
@@ -0,0 +1,202 @@
+       * 3 averaged c axis of olivine -> 3 (dim) doubles, starts at:
+       *                                   data_position + 7
+       * 4 averaged a axis of enstatite -> 3 (dim) doubles, starts at:
+       *                                    data_position + 10
+       * 5 averaged b axis of enstatite -> 3 (dim) doubles, starts at:
+       *                                    data_position + 13
+       * 6 averaged c axis of enstatite -> 3 (dim) doubles, starts at:
+       *                                    data_position + 16
+       *
+       * @ingroup ParticleProperties
+       */
+      template <int dim>
+      class CpoBinghamAverage : public Interface<dim>, public ::aspect::SimulatorAccess<dim>
+      {
+        public:
+          /**
+           * constructor
+           */
+          CpoBinghamAverage();
+
+          /**
+           * Initialization function. This function is called once at the
+           * beginning of the program after parse_parameters is run.
+           */
+          virtual
+          void
+          initialize ();
+
+          /**
+           * Initialization function. This function is called once at the
+           * creation of every particle for every property to initialize its
+           * value.
+           *
+           * @param [in] position The current particle position.
+           * @param [in,out] particle_properties The properties of the particle
+           * that is initialized within the call of this function. The purpose
+           * of this function should be to extend this vector by a number of
+           * properties.
+           */
+          virtual
+          void
+          initialize_one_particle_property (const Point<dim> &position,
+                                            std::vector<double> &particle_properties) const;
+
+          /**
+           * Update function. This function is called every time an update is
+           * request by need_update() for every particle for every property.
+           *
+           * @param [in] data_position An unsigned integer that denotes which
+           * component of the particle property vector is associated with the
+           * current property. For properties that own several components it
+           * denotes the first component of this property, all other components
+           * fill consecutive entries in the @p particle_properties vector.
+           *
+           * @param [in] position The current particle position.
+           *
+           * @param [in] solution The values of the solution variables at the
+           * current particle position.
+           *
+           * @param [in] gradients The gradients of the solution variables at
+           * the current particle position.
+           *
+           * @param [in,out] particle_properties The properties of the particle
+           * that is updated within the call of this function.
+           */
+          virtual
```

Comment:
as above, and for all other functions in this header that are marked as `virtual`